### ISSUE TITLE
Support for German accounts? Bot does not recognize accounts with full points.

### ISSUE THREAD

**Issue by MacMarde** [xMNG/Microsoft-Rewards-Bot#48]
Hi I am new here. I really appreciate your bot and your work.

I am trying to use the bot with my German accounts. I have set the website language to US English and I think the website does now look like with a real US account.

But I am also having the problem that the bot doesnt recognize accounts with full points. (Most probably the bot also does not work corrctly with empty points. But as for today all my accounts have full points)
I just started an account with full points. I can see in the log that the bot made 152 mobile searches, then started a new HTTP connection for making the PC searches. It seems like for every PC search I am now getting the following console error:
"ERROR:web_contents_delegate.cc(197)] WebContentsDelegate::CheckMediaAccessPermission: Not supported. "

I am willing to contribute to the project as much as I can. Please help me to make it work with my german accounts. I do know a little bit about coding. Maybe we can work together to implement support for german accounts.

Now Im also got this console error for some PC searches (I started from the Command prompt with administartor rights):
[0504/075749.009:INFO:CONSOLE(57)] "Uncaught SecurityError: Failed to set the 'href' property on 'Location': The current window does not have permission to navigate the target frame to 'https://login.microsoftonline.com/common/oauth2/authorize?client_id=9ea1ad79-fdb6-4f9a-8bc3-2b70f96e34c7&response_type=id_token&nonce=ecf2dfbb-7174-43f0-a7d5-4ebea31a4cbc&redirect_uri=https%3a%2f%2fwww.bing.com%2forgid%2fidtoken%2fnosignin&scope=openid&response_mode=form_post&msafed=0&prompt=none&iframe-request-id=4c4e5830-6d49-4acf-9b3c-58e561180900'.", source: https://login.microsoftonline.com/common/oauth2/authorize?client_id=9ea1ad79-fdb6-4f9a-8bc3-2b70f96e34c7&response_type=id_token&nonce=ecf2dfbb-7174-43f0-a7d5-4ebea31a4cbc&redirect_uri=https%3a%2f%2fwww.bing.com%2forgid%2fidtoken%2fnosignin&scope=openid&response_mode=form_post&msafed=0&prompt=none (57)"

It made over 100 PC searches so far. I think Im gonna stop it now.

**Comment by ShoGinn** [comment]
Language shouldn't matter, but according to your description it does?

Can you give examples of the name changes?

**Comment by MacMarde** [comment]
No I cant give examples of the name changes. I dont know how to find them. Maybe if you assist me I will try to find the name changes. I have never used Selenium.
I am also not sure that this is the problem at all, it was just a quick shot.

How can we figure out why the bot does not recognize the points from the dashboard?
I tried to observe it without headless mode. It seems for me like the dashboard wont open at all, instead after the login it loads the landing site. 
Then it tries to open the dashboard site, which I can only recognize by some 100ms in the browser url. The site seems not to load succesfull, instead it instantly redirects to the rewards site.

I tried it manually with the bot URL "https://account.microsoft.com/rewards/dashboard". And actually this site instantly redirects to "https://account.microsoft.com/rewards/"

If I want to see my Points breakdown I need to click on "Points breakdown" in the upper right corner of the rewards website, which leads to this site: "https://account.microsoft.com/rewards/pointsbreakdown"

Here I can see my points with German text (unlike the rest of the website). 
So basically it seem like the dashboard is different with german accounts, or at least has a different URL.

What do you think @ShoGinn ?

If there is anything I can support you with to solve the problem just ask.

**Comment by MacMarde** [comment]
I just tried replcaing the URL in the bot to 
DASHBOARD_URL = 'https://account.microsoft.com/rewards/pointsbreakdown' 
But its not working. It seems to be more complicated like that. Any way you can reproduce it or can I assist you with that?

**Comment by ShoGinn** [comment]
This is what it looks like in the us:
![image](https://user-images.githubusercontent.com/2660603/57183357-b25f0c00-6e60-11e9-92ad-4880d85d71aa.png)

But it pulls the progress at the chrome extension url:
https://www.bing.com/rewardsapp/bepflyoutpage?style=chromeextension

And that looks like this:
![image](https://user-images.githubusercontent.com/2660603/57183382-18e42a00-6e61-11e9-97d1-0fa869aca2ea.png)

**Comment by MacMarde** [comment]
Im not sure if I do understand you.
You mean this is what it looks like for https://account.microsoft.com/rewards/pointsbreakdown, 
or is this what it looks like for "https://account.microsoft.com/rewards/dashboard ?

Anyway I also dont understand what you mean with:?

> "But it pulls the progress at the chrome extension url:
> https://www.bing.com/rewardsapp/bepflyoutpage?style=chromeextension"

Do you mean you get redirected to that page showing the gift card?

With german account it looks like this for https://account.microsoft.com/rewards/pointsbreakdown and it stays on this site:

![grafik](https://user-images.githubusercontent.com/50266741/57183520-9c783800-6eae-11e9-8800-1e50ce7b6018.png)

It seems these both versions are very similar. Any ideas what to do next to fix the issue?

**Comment by ShoGinn** [comment]
The chrome extension is used to check when to stop pc searches.

what does the screen show for: https://www.bing.com/rewardsapp/bepflyoutpage?style=chromeextension

**Comment by MacMarde** [comment]
It looks like this:

![grafik](https://user-images.githubusercontent.com/50266741/57183679-c4689b00-6eb0-11e9-854a-fdfcf5adf8d8.png)

**Comment by MacMarde** [comment]
Now its getting weird. The whole situation changed.

The above image from the URL https://www.bing.com/rewardsapp/bepflyoutpage?style=chromeextension was posted from the bot "Microsoft Rewards automater" which was taken down by MS recently. In this bot I used the function to manually open the browser for my account. 

But now I opened my account without the bot. I just started chrome manually and logged in manually. Now I do have a completely other screen for https://www.bing.com/rewardsapp/bepflyoutpage?style=chromeextension :

![grafik](https://user-images.githubusercontent.com/50266741/57184581-d270e880-6ebd-11e9-9fe6-9bfabcc2224d.png)

I am still logged in. Maybe this is the reason why the bot cant find the elements. For some reason the content changed. If I do let the python bot run maybe its the same problem. I can not confirm it, because the site is not displayed correctly and only for some milliseconds with the python bot.

**Comment by MacMarde** [comment]
Ok I cant continue.
I see two possibilities:

1. Find a way to change the chrome browser so that the site https://www.bing.com/rewardsapp/bepflyoutpage?style=chromeextension displays the points, or
2. Use the site https://account.microsoft.com/rewards/pointsbreakdown for point extraction. 
But I dont know how I do extract those information to use with the "browser.find_element_by_class_name" method. Im sure there are tools or tricks how to extract those informations. I would appreciate any help. 

I hope someone can help me?

**Comment by MacMarde** [comment]
Ok I have solved it so far. Im proud of myself because until today I have never used Python or Selenium or HTML.

I used the second solution, which is to use the site  "https://account.microsoft.com/rewards/pointsbreakdown" to extract point information.
This site has no information about total points so I just commented it out in the source code, but this isnt important for now. This information can be obtained from the dashboard later.
If this solution is also working for US accounts, we should think about merging my code. Maybe its more compatible for international support. So could someone please test it with US account?
If its not working with US account we could still think about implement it as an region option.

I have changed the code and used the "browser.find_element_by_xpath()" method, because the element is not clearly identificable by the class name. Therefore I used the XPath Helper Addon. The resulting string looks like hell but its working. Only problem I do see is that it is not stable against small website changes. If someone with more experience knows a better way feel free to let me know or change it yourself.


[ms_rewards.zip](https://github.com/xMNG/Microsoft-Rewards-Bot/files/3144945/ms_rewards.zip)

**Comment by ShoGinn** [comment]
@MacMarde I'll take a look and see.. for future reference, if you fork the repo and then make changes and create a pull request you can reference it!

**Comment by MacMarde** [comment]
ok I will try it next time. Not sure if I make it right.

**Comment by MacMarde** [comment]
I did let the bot run today and it worked perfect. I get a couple of error messages in the consoe, but it seems not to be a problem. I open a new topic about this error messages.
@ShoGinn Please let me if its also working for you.

**Comment by MacMarde** [comment]
Bot crashes on LVL 1 accounts, becasue the element for "mobile searches" does not exist. 
Dont know how to get around this. Maybe a simple try and expetion statement around the XPATH in the code?

try:
current_mobile_points, max_mobile_points = map(
            int, browser.find_element_by_xpath(....)
except no_mobile_searches_available:
           current_mobile_points = 0
           max_mobile_points = 0

**Comment by ShoGinn** [comment]
Lvl1 accounts are a pain to fix. 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by MacMarde** [comment]
ok I have fixed the code. How should I upload it?

**Comment by MacMarde** [comment]
@ShoGinn My suggestion is now to make those code changes mentioned above to my local github desktop repo and then commit them? And then pull them or first a pull request?
With this changes the bot is working for german regions with both LVL 1 and LVL 2.
It would be a good practice and I can learn how to contribute to the code in the future.
Only one question? To which branch should I add the code changes? Master or dev?

**Comment by ShoGinn** [comment]
Currently you are asking for changes on the master branch.
I unfortunately do not have time to take on an apprentice ;)

Your goal is to create a pull request to suggest changes with your fixes.

Reference this issue when you do the pull request.

**Comment by xMNG** [comment]
@ShoGinn 
Thanks for all your hard work! I know you are busy, but would you be willing to take me on as apprentice/mentee? I'm currently back in school.

**Comment by ShoGinn** [comment]
>@ShoGinn  ...

@xMNG you have already surpassed me ;) don’t fall into my bad habits 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by MacMarde** [comment]
@ShoGinn I will try to implement some code changes and make a pull request in a few days. Also for the other issues regarding optimized number of searches with region parameter and maybe also a user defineable searchterm list.

**Comment by ShoGinn** [comment]
Im curious if changing to English fixes this localization problem?

**Comment by MacMarde** [comment]
@ShoGinn I already told you I changed the site language to English. It is not working.
I will make a Push request later this week with my suggested code changes. They do work. You already also do have a copy of this.

**Comment by MacMarde** [comment]
Anyone knows how to rech this site manually from the menu without URL?

 https://www.bing.com/rewardsapp/bepflyoutpage?style=chromeextension

Because maybe this way I can figure out why its not working for my accounts.

**Comment by ShoGinn** [comment]
The url was pulled from the chrome extension. Getting to it without the url is not possible 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by MacMarde** [comment]
Ok so where to get the extension to test it?

**Comment by MacMarde** [comment]
I have optimized my code, comitted it to github desktop and will make a push request after I have tested it some days.

**Comment by MacMarde** [comment]
@ShoGinn ok I have made a pull request

**Comment by MacMarde** [comment]
It seems like my solution is not working for UK accounts. See #71 
Maybe I will add a region option on command line, so that german accounts will be working too. I dont know a better solution. What do you think?

The "Microsoft rewards automator" which was available recently uses the following URL to extract those information and is working for german accounts too:
https://account.microsoft.com/rewards/api/getuserinfo?X-Requested-With=XMLHttpRequest

It seems like it builds a C# dictionary with all keys and values from this URL using javascript. Then it extracts all information needed. But I dont really understand it. If someone wants the sourcecode I can share it.

**Comment by ShoGinn** [comment]
I have decided to cease development on this project to avoid any legal issues seen by other developers. Not worth it to continue to develop a bot from my perspective. Sorry @xMNG

**Comment by MacMarde** [comment]
@ShoGinn What did change your mind? Which other developers do you mean?

**Comment by abdo44705** [comment]
Hello MacMarde can you please share with me The "Microsoft rewards automator" which was available on Reddit cuz this python  bot doesn't work in my region .this my gp1849769@gmail.com thanks

**Comment by MacMarde** [comment]
No sorry. The bot stopped working a long time ago. It is not useful.

**Comment by abdo44705** [comment]
yeah I know cuz I was using it but about a week ago the bot stopped
working  it gives me an error { dashboard  unavailable} since I don't have
coding I tried to search the internet to find how to fix it.then I
found your thread on gethub

**Comment by abdo44705** [comment]
the bot now can't extract dashboard information using this url
https://account.microsoft.com/rewards/api/getuserinfo?X-Requested-With=XMLHttpRequest
.

**Comment by abdo44705** [comment]
https://account.microsoft.com/rewards/api/getuserinfo?X-Requested-With=XMLHttpRequest
.
This url is not working anymore. can you please help me to fix it

**Comment by MacMarde** [comment]
Sorry I can't help you. I stopped contributing to this project.

**Comment by abdo44705** [comment]
ok thank you anyway
